_posts/2023-09-05-wallclock-profiler.md

```diff
@@ -0,0 +1,219 @@
+
+So async-profiler's excellent codebase gave us a starting point for implementing some of the features we had in mind.
+
+Below is a flamegraph of the same application as above recorded with Datadog's customisation of async-profiler. 
+Notice the ZGC activity to the left and compare it to the GC activity we could infer from JFR above, there's more detail and it's more obvious that the service has a GC problem.
+
+![async-profiler all activity](/assets/2023/09/wallclock-profiler/datadog-flamegraph-cpu-time-all-activity.png)
+
+Having access to the source code made it possible to add new fields to event types, such as correlation identifiers.
+Decorating samples with span ids makes it possible to link each sample to the endpoint being served, rather than guess based on the frame name (which only gets harder outside of contrived examples).
+This makes it possible to filter the flamegraph by endpoint instead of guessing based on the frame names:
+
+![async-profiler all activity](/assets/2023/09/wallclock-profiler/datadog-flamegraph-cpu-time-endpoint-filter.png)
+
+The green frames to the left, which are JNI calls into the lz4-java library, are missing from the JFR flamegraph because JFR won't unwind stacks when the JVM is executing native code.
+In this particular case, this doesn't actually distort the flamegraph _much_, because LZ4 compression is very fast and doesn't need as much CPU time as the random number generation we already know about, but if it were the other way around the JFR flamegraph would be very misleading. 
+
+Even though the flamegraph now shows a complete picture of CPU time, it still points towards optimising random number generation, which we can't do.
+The problem is now that CPU time isn't always enough to understand latency.
+
+## Wallclock profiling
+
+As used in Datadog's profiler, async-profiler's CPU profiler uses CPU time to schedule sampling, which means time spent off-CPU will not be sampled. 
+In Java `Thread` terms, a CPU profiler can only sample threads in the `RUNNABLE` state, and ignores threads in the `BLOCKED`, `TIMED_WAITING`, and `WAITING` states, but a _wallclock profiler_ can sample threads in any state.
+async-profiler also has a wallclock profiler, but we found we needed to tailor it to our requirements.
+
+### Sampling strategy
+
+The first issue we encountered was with overhead (for our own particular requirements, don't be discouraged from using async-profiler's wallclock profiler in a different context).
+CPU profiling is always very lightweight because the overhead is proportional to the number of cores, whereas for wallclock profiling, overhead scales with the number of threads.
+async-profiler's wallclock profiler samples sets of up to 8 threads round-robin, and to keep the sampling interval consistent as the number of threads changes, will adjust the sampling interval to as low 100 microseconds.
```

Comment:
```suggestion
async-profiler's wallclock profiler sampler sets of up to 8 threads round-robin, and to keep the sampling interval consistent as the number of threads changes, will adjust the sampling interval to as low 100 microseconds.
```

Reply:
no, it _samples_ sets of threads. I'll rework the sentence to make this easier to read though.